Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,232 @@
+      const std::shared_ptr<BooleanModel>& model, const Type& value) {
+    auto inserted_model = m_models.insert({value, model});
+    if(!inserted_model.second) {
+      return;
+    }
+    if(model->get_current()) {
+      model->set_current(false);
+    }
+    m_connections.push_back(model->connect_current_signal(
+      [=] (auto is_selected) { on_current(value, is_selected); }));
+    if(get_state() == QValidator::Invalid) {
+      set_current(value);
+    }
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::make_association(
+      const Type& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::find(
+      const Type& value) const {
+    auto iterator = m_models.find(value);
+    if(iterator == m_models.end()) {
+      return nullptr;
+    }
+    return iterator->second;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::get_state() const {
+    if(m_models.find(m_current) != m_models.end()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template<typename T>
+  const typename AssociativeValueModel<T>::Type&
+      AssociativeValueModel<T>::get_current() const {
+    return m_current;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::set_current(const Type& value) {
+    if(m_models.find(value) == m_models.end()) {
+      return QValidator::Invalid;
+    }
+    auto previous = m_current;
+    m_current = value;
+    set_associated_model_value(previous, false);
+    set_associated_model_value(m_current, true);
+    m_current_signal(m_current);
+    return QValidator::Acceptable;
+  }
+
+  template<typename T>
+  boost::signals2::connection AssociativeValueModel<T>::connect_current_signal(
+      const typename CurrentSignal::slot_type& slot) const {
+    return m_current_signal.connect(slot);
+  }
+
+  template<typename T>
+  void AssociativeValueModel<T>::set_associated_model_value(const Type& value,
+      bool model_value) {
+    if(auto iterator = m_models.find(value); iterator != m_models.end()) {
+      iterator->second->set_current(model_value);
+    }
+  }
+
+  template<typename T>
+  void AssociativeValueModel<T>::on_current(const Type& value,
+      bool is_selected) {
+    if(value == m_current && is_selected ||
+        value != m_current && !is_selected) {
+      return;
+    }
+    if(m_is_blocked) {
+      if(is_selected) {
+        set_associated_model_value(value, false);
+      }
+      return;
+    }
+    if(m_current == value && !is_selected) {
+      if(m_default_value) {
+        m_is_blocked = true;
+        set_current(*m_default_value);
+        m_is_blocked = false;
+      } else {
+        set_associated_model_value(m_current, true);
+      }
+      return;
+    }
+    m_is_blocked = true;
+    set_current(value);
```

Comment:
As long as set_current(...) isn't called again in response to changing a model's value, there shouldn't be an issue, and the m_is_blocked variable prevents that from happening.

What conditions would cause a bug here?

Reply:
```m_is_blocked``` prevents what from happening?

Reply:
m_is_blocked is what prevents set_current from being called again. From my understanding if set_current is called and the current signal is emitted again it could result in problems with reentrancy, but in on_current I check the blocked state of the associative model before calling set_current to ensure that any change in a model doesn't result in infinitely recursive model updates caused by the current signal being emitted, which updates one of the associated models, which updates the previous associated model, and so on.

The goal is that any model becoming true should result in one and only one current value being signaled, so if emitting the current signal results in another model being set to true, the second model's value is reverted to false and the original change is the one that is kept.

A potential issue I can see is that something could be holding a pointer to an associated model and could potentially also be reverting the value, resulting in both the AssociativeValueModel and a component holding a pointer to the associated BooleanModel having a 'tug-of-war' of sorts as each model reverts the value, but I'm not sure there's a solution for that in the AssociativeValueModel itself.

Reply:
So then in this case it not about preventing infinite loops but ensuring internal consistency? A signal handler could call set_current(...) or manipulate a model (resulting in a change to current), and it's always accepted?

Reply:
If the value isn't accepted then ```set_current``` returns ```Invalid```, the issue I have is with re-entrancy, not with infinite loops.

Reply:
Before I commit to this and rewrite the tests, I wanted to confirm this was the idea: https://gist.github.com/spireken/22620313c805d07e23b5fd393c07d98b

The operation of set_current can be summed up to:

- If the value is not associated, reject it and return Invalid.
- If the model is in the middle of setting it's current value then it's considered 'blocked', and any subsequent calls to set_current will be put into a queue. This way, everything listening to the current signal has a chance to see every change.
- After signalling the change of current, if the queue isn't empty, then recursively call set_current with the value at the front of the queue, until it's empty.
- A call to to set_current while it's blocked returns Intermediate, meaning that the value was acknowledged as valid but is not immediately set as the current.

The model only changes the associated models as required to implement the exclusive behavior regarding the current value, so it would be possible to 'break' it in some ways by manipulating the associated models using some external component.

Reply:
The model should never be in the middle of setting its value. Only signal at the very end of the operation, never in the middle of the operation.

Reply:
That's what it does, but one value at a time. By the time the current value is signaled the model has that value as the current value, but in the case where something has set another current value in response the given value is queued for later.

In short, the intent is that in no case should a signal be emitted for a value that isn't the current. In all cases calling get_current() in response to a current signal will always be equal to the value emitted with the signal, no matter how many times set_current was called in response to the current signal.

The test case you gave for reentrant signals passes with this implementation, because it ensures every listener receives the initial current change before receiving any additional current change signals from a reentrant call to set_current.

Reply:
Okay good, push the changes.